### PR TITLE
FIX: Add getters for camera parameters #1419

### PR DESCRIPTION
**Pull request by Schwarzemann**
I added two new methods named `rotateByMouse(double deltaX, double deltaY)` and `rotateAroundFocalPoint(double deltaX, double deltaY)`. The `rotateByMouse` method calculates calculates pitch and yaw from mouse variables. The `rotateAroundFocalPoint` calculates by rotating the camera around focal point.

### CODE REVIEW

File: library/public/camera.h [hunk 1]
```diff
@@ -73,6 +73,9 @@ class F3D_EXPORT camera
 
   ///@}
 
+  virtual camera& rotateByMouse(double deltaX, double deltaY) = 0;
+  virtual camera& rotateAroundFocalPoint(double deltaX, double deltaY) = 0;
```
Comment (mwestphal) [inline library/public/camera.h:77]: I'm not sure to follow how that relates to #1419

Comment (Schwarzemann) [reply library/public/camera.h:77]: That's also what I realized. This was just a feature that seemed cool to me. I will implement something later.

Comment (mwestphal) [reply library/public/camera.h:77]: That's ok,  you can add new features if you want, but please discuss them in discord/ a dedicated issue first :)

Comment (Schwarzemann) [reply library/public/camera.h:77]: Yeah my bad sorry. I was working on this separetely but PRs got mixed. I will separate them tonight.

File: library/public/camera.h [hunk 2]
```diff
@@ -103,4 +106,4 @@ class F3D_EXPORT camera
 };
 }
 
-#endif
+#endif
```
Comment (mwestphal) [inline library/public/camera.h:109]: missing end of line

File: library/public/camera.h [hunk 3]
```diff
@@ -49,6 +49,9 @@ class F3D_EXPORT camera
   virtual camera& setState(const camera_state_t& state) = 0;
   virtual camera_state_t getState() = 0;
   virtual void getState(camera_state_t& state) = 0;
+  virtual angle_deg_t getAzimuth() = 0;
+  virtual angle_deg_t getYaw() = 0;
+  virtual angle_deg_t getElevation() = 0;
```
Comment (mwestphal) [inline library/public/camera.h:54]: put that in another doxygen block and add some doc

File: library/private/camera_impl.h [hunk 4]
```diff
@@ -50,6 +50,9 @@ class camera_impl : public camera
   camera& setState(const camera_state_t& state) override;
   camera_state_t getState() override;
   void getState(camera_state_t& state) override;
+  angle_deg_t getAzimuth() override;
+  angle_deg_t getYaw() override;
+  angle_deg_t getElevation() override;
```
Comment (mwestphal) [inline library/private/camera_impl.h:55]: @snoyer maybe getCurrentElevation and such would be better ?

Comment (mwestphal) [reply library/private/camera_impl.h:55]: or even computeElevation ?

File: library/public/camera.h [hunk 5]
```diff
@@ -49,6 +49,9 @@ class F3D_EXPORT camera
   virtual camera& setState(const camera_state_t& state) = 0;
   virtual camera_state_t getState() = 0;
   virtual void getState(camera_state_t& state) = 0;
+  virtual angle_deg_t getAzimuth() = 0;
```
Comment (Meakk) [inline library/public/camera.h:52]: we need to add Python bindings

File: library/src/camera_impl.cxx [hunk 6]
```diff
@@ -117,6 +117,60 @@ angle_deg_t camera_impl::getViewAngle()
...
+  double viewDir[3];
+  vtkMath::Subtract(foc, pos, viewDir);
+  double viewDirProj[3] = { viewDir[0], viewDir[1], 0.0 };
+  double norm = vtkMath::Norm(viewDirProj);
+  if (norm < VTK_DBL_EPSILON)
+  {
+    return 0.0;
+  }
+  vtkMath::Normalize(viewDirProj);
+  return vtkMath::DegreesFromRadians(atan2(viewDirProj[1], viewDirProj[0]));
```
Comment (Meakk) [inline library/src/camera_impl.cxx:136]: Looking at this page for reference https://en.wikipedia.org/wiki/List_of_common_coordinate_transformations#To_spherical_coordinates
While it's important to check that the norm is not zero, I think you can skip the normalization.
You can also check the squared normalization only to save a sqrt for performance reason.
Here's a suggestion (please check it works)

```suggestion
  double viewDirProj[2] = { viewDir[0], viewDir[1] };
  if (vtkMath::SquaredNorm<2>(viewDirProj) < VTK_DBL_EPSILON)
  {
    return 0.0;
  }
  return vtkMath::DegreesFromRadians(atan2(viewDirProj[1], viewDirProj[0]));
```

File: library/src/camera_impl.cxx [hunk 7]
```diff
@@ -117,6 +117,60 @@ angle_deg_t camera_impl::getViewAngle()
...
+  cam->GetFocalPoint(foc);
+  double viewDir[3];
+  vtkMath::Subtract(foc, pos, viewDir);
+  double viewDirProj[3] = { viewDir[0], 0.0, viewDir[2] };
+  double norm = vtkMath::Norm(viewDirProj);
+  if (norm < VTK_DBL_EPSILON)
+  {
+    return 0.0;
+  }
+  vtkMath::Normalize(viewDirProj);
```
Comment (Meakk) [inline library/src/camera_impl.cxx:154]: same comment here.

File: library/private/camera_impl.h (hunk 4 above)
Comment (Schwarzemann) [reply library/private/camera_impl.h:55]: I changed the naming convention

File: library/public/camera.h (hunk 5 above)
Comment (Schwarzemann) [reply library/public/camera.h:52]: Python bindings are added.

File: library/src/camera_impl.cxx (hunk 6 above)
Comment (Schwarzemann) [reply library/src/camera_impl.cxx]: I applied the changes you suggested.

File: library/src/camera_impl.cxx (hunk 7 above)
Comment (Schwarzemann) [reply library/src/camera_impl.cxx]: Same here.

File: library/public/camera.h (hunk 3 above)
Comment (Schwarzemann) [reply library/public/camera.h:60]: Added

File: java/F3DJavaBindings.cxx [hunk 8]
```diff
@@ -138,17 +138,17 @@ extern "C"
 
   JNIEXPORT void JAVA_BIND(Camera, azimuth)(JNIEnv* env, jobject self, jdouble angle)
   {
-    GetEngine(env, self)->getWindow().getCamera().azimuth(angle);
+    GetEngine(env, self)->getWindow().getCamera().addAzimuth(angle);
```
Comment (Meakk) [inline java/F3DJavaBindings.cxx:141]: Please change the api name in the JAVA_BIND macro

File: library/public/camera.h (hunk 2 above)
Comment (Meakk) [reply library/public/camera.h:115]: Still missing empty line

File: python/F3DPythonBindings.cxx [hunk 9]
```diff
@@ -268,9 +268,12 @@ PYBIND11_MODULE(pyf3d, module)
     .def("pan", &f3d::camera::pan, py::arg("right"), py::arg("up"), py::arg("forward") = 0.0)
     .def("zoom", &f3d::camera::zoom)
     .def("roll", &f3d::camera::roll)
-    .def("azimuth", &f3d::camera::azimuth)
-    .def("yaw", &f3d::camera::yaw)
-    .def("elevation", &f3d::camera::elevation)
+    .def("azimuth", &f3d::camera::addAzimuth)
```
Comment (Meakk) [inline python/F3DPythonBindings.cxx:271]: Let's change the python api name to add_azimuth too

File: java/F3DJavaBindings.cxx (hunk 8 above)
Comment (Schwarzemann) [reply java/F3DJavaBindings.cxx:141]: Changed.

File: python/F3DPythonBindings.cxx (hunk 9 above)
Comment (Schwarzemann) [reply python/F3DPythonBindings.cxx]: Changed.

File: python/F3DPythonBindings.cxx [hunk 10]
```diff
@@ -268,9 +268,12 @@ PYBIND11_MODULE(pyf3d, module)
...
     .def("roll", &f3d::camera::roll)
-    .def("azimuth", &f3d::camera::azimuth)
-    .def("yaw", &f3d::camera::yaw)
-    .def("elevation", &f3d::camera::elevation)
+    .def("add_azimuth", &f3d::camera::addAzimuth)
+    .def("add_yaw", &f3d::camera::addYaw)
+    .def("add_elevation", &f3d::camera::addElevation)
+    .def("get_azimuth", &f3d::camera::getAzimuth)
+    .def("get_yaw", &f3d::camera::getYaw)
+    .def("get_elevation", &f3d::camera::getElevation)
```
Comment (snoyer) [inline python/F3DPythonBindings.cxx:276]: that could be properties, see `image` class for examples
```suggestion
    .def_property_readonly("azimuth", &f3d::camera::getAzimuth)
    .def_property_readonly("yaw", &f3d::camera::getYaw)
    .def_property_readonly("elevation", &f3d::camera::getElevation)
```

Comment (Schwarzemann) [reply python/F3DPythonBindings.cxx]: Implemented.

File: library/public/camera.h (hunk 2 above)
Comment (mwestphal) [reply library/public/camera.h:115]: still missing

File: library/src/camera_impl.cxx [hunk 11]
```diff
@@ -117,6 +117,55 @@ angle_deg_t camera_impl::getViewAngle()
...
+  vtkCamera* cam = this->GetVTKCamera();
+  double pos[3], foc[3];
+  cam->GetPosition(pos);
+  cam->GetFocalPoint(foc);
+  double viewDir[3];
+  vtkMath::Subtract(foc, pos, viewDir);
+  double viewDirProj[2] = { viewDir[0], viewDir[1] };
+  if (vtkMath::Dot2D(viewDirProj, viewDirProj) < VTK_DBL_EPSILON)
+  {
+    return 0.0;
```
Comment (mwestphal) [inline library/src/camera_impl.cxx:132]: please add testing for this

File: library/src/camera_impl.cxx [hunk 12]
```diff
@@ -117,6 +117,55 @@ angle_deg_t camera_impl::getViewAngle()
...
+  vtkCamera* cam = this->GetVTKCamera();
+  double pos[3], foc[3];
+  cam->GetPosition(pos);
+  cam->GetFocalPoint(foc);
+  double viewDir[3];
+  vtkMath::Subtract(foc, pos, viewDir);
+  double viewDirProj[2] = { viewDir[0], viewDir[2] };
+  if (vtkMath::Dot2D(viewDirProj, viewDirProj) < VTK_DBL_EPSILON)
+  {
+    return 0.0;
```
Comment (mwestphal) [inline library/src/camera_impl.cxx:149]: please add testing for this

File: library/src/camera_impl.cxx [hunk 13]
```diff
@@ -163,7 +212,6 @@ camera& camera_impl::dolly(double val)
   this->Internals->VTKRenderer->ResetCameraClippingRange();
   return *this;
 }
-
```
Comment (mwestphal) [inline library/src/camera_impl.cxx:166]: please keeop this line

File: python/testing/test_camera.py [hunk 14]
```diff
@@ -80,9 +80,9 @@ def test_moves():
     camera.dolly(10)
     angle = 30
     camera.roll(angle)
-    camera.azimuth(angle)
-    camera.yaw(angle)
-    camera.elevation(angle)
+    camera.add_azimuth(angle)
+    camera.add_yaw(angle)
+    camera.add_elevation(angle)
```
Comment (mwestphal) [inline python/testing/test_camera.py:85]: add testing for the new properties

**Review by mwestphal — CHANGES_REQUESTED** [review]
small changes needed

**Comment by mwestphal** [comment]
do you need any help with branch management ? Should I reopen this ?

**Comment by Schwarzemann** [comment]
> do you need any help with branch management ? Should I reopen this ?

Ah no worries. I created a different branch for testing aside from master. I'll reopen this PR when ready.

**Pull request closed without merging**